docs/rules/no-implicit-injections.md

```diff
@@ -0,0 +1,115 @@
+
+<!-- end auto-generated rule header -->
+
+Ember 3.26 introduced a deprecation for relying on implicit service injections or allowing addons to implicitly inject services into all classes of certain types. Support for this is dropped in Ember 4.0.
+
+In many applications, `this.store` from Ember Data is often used without injecting the `store` service in Controllers or Routes. Other addons may also have included implicit service injections via initializers and the `application.inject` API.
+
+To resolve this deprecation, a service should be explicitly declared and injected using the [service injection decorator](https://api.emberjs.com/ember/3.28/functions/@ember%2Fservice/inject).
+
+## Rule Details
+
+This rule checks for a configured list of previously auto injected services and warns if they are used in classes without explicit injected service properties.
+
+## Examples
+
+Examples of **incorrect** code for this rule:
+
+```js
+// routes/index.js
+import Route from '@ember/routing/route';
+
+export default class IndexRoute extends Route {
+  model() {
+    return this.store.findAll('rental');
+  }
+}
+
+```
+
+```js
+// controllers/index.js
+import Controller from '@ember/controller';
+import { action } from '@ember/object';
+
+export default class IndexController extends Controller {
+  @action
+  loadUsers() {
+    return this.store.findAll('user');
+  }
+}
+```
+
+Examples of **correct** code for this rule:
+
+```js
+// routes/index.js
+import Route from '@ember/routing/route';
+import { inject as service } from '@ember/service';
+
+export default class IndexRoute extends Route {
+  @service('store') store;
+
+  model() {
+    return this.store.findAll('rental');
+  }
+}
+```
+
+```js
+// controller/index.js
+import Route from '@ember/routing/route';
+import { action } from '@ember/object';
+import { inject as service } from '@ember/service';
+
+export default class IndexController extends Controller {
+  @service('store') store;
+
+  @action
+  loadUsers() {
+    return this.store.findAll('user');
+  }
+}
+```
+
+## Migration
+
+The autofixer for this rule will update classes and add injections for the configured services.
+
+## Configuration
+
+This lint rule will search for instances of `store` used in routes or controllers by default. If you have other services that you would like to check for uses of, the configuration can be overridden.
+
+- object -- containing the following properties:
+  - array -- `services` -- Array of configuration objects configuring the lint rule to check for use of implicit injected services
+    - string -- `serviceName` -- The name of the service that is implicitly injected
+    - array -- `moduleNames` -- Array of string listing the types of classes (controller, route, component, etc) to check for implicit injections. If an array is declared only those class types will be checked for implicit injection. (Defaults to checking all class files/types)
```

Comment:
Regarding `moduleNames`, how critical do you think this option is?

Just to explore an alternative design, it's possible we could use ESLint [overrides](https://eslint.org/docs/latest/user-guide/configuring/configuration-files#configuration-based-on-glob-patterns) to achieve mostly the same thing but with a much simpler version of the rule:

```js
// .eslintrc.js
module.exports = {
  overrides: [
    {
      files: ['controllers/**/*.js'],
      rules: {
        'ember/no-implicit-injections': ['error', { services: ['store'] }],
      },
    },
    {
      files: ['components/**/*.js', 'routes/**/*.js'],
      rules: {
        'ember/no-implicit-injections': ['error', { services: ['someOtherService'] }],
      },
    },
  ],
};
```

The user can use this built-in feature of ESLint to easily apply the rule to whatever type of file they prefer, with the full customizability that these glob file patterns provide.

And most of the time, the user won't even need to target specific types of Ember classes anyway, right? As in, I assume most of the time the default of "checking all class files/types" is sufficient. So we could reduce the complexity of the rule a lot if we don't need to worry about checking for specific Ember classes.

If you think `moduleNames` is still important, let us know why. I just want to make sure we don't add unnecessary complexity that can already be achieved with `overrides`, which is a mistake I've made myself before.

Reply:
Overrides on files doesn't properly account for class types, pod structures, or the way that services are injected by the Ember initializers.

The `moduleNames` (maybe `moduleTypes` would be a better property name for it) allows for better identification since it uses `isEmberCoreModule` for module type identification to better know what kind of module is being checked from the base class.

Reply:
For modules with more than one class (pod or just native classes) this could also become overly zealous.

Ex a team exports route and controller in a single module. Implicit uses of `media` are allowed in routes but not in controllers. This should be configurable and file based overrides will not account for this.

A more common example would be that a component declares the component *and* a native class. The native class is not subject to the implicit service injection rule.

---

I'll add test cases for both of these.

Reply:
Test cases added

Reply:
Makes sense, thanks for the justification.